llumnix/backends/vllm/migrate_backend.py

```diff
@@ -0,0 +1,257 @@
+                        .format(self.group_name, self.backend))
+            self.group_name = None
+
+    def migrate_cache(self, src_handle, src_blocks: List[int], dst_blocks: List[int]) -> None:
+        tot_blocks = len(src_blocks)
+        src_rank = ray.get(self.actor.exec_method.remote(self.is_driver_worker, src_handle, "get_ray_rank"))
+
+        for start_idx in range(0, tot_blocks, self.num_migration_cache_blocks):
+            offset = min(self.num_migration_cache_blocks, tot_blocks - start_idx)
+            send_blocks = src_blocks[start_idx:start_idx+offset]
+            recv_blocks = dst_blocks[start_idx:start_idx+offset]
+            self.actor.exec_method.remote(self.is_driver_worker, src_handle, "do_send", self.ray_rank, send_blocks)
+            self.do_recv(src_rank, recv_blocks)
+
+    def do_send(self, dst_handle, blocks: List[int]):
+        num_blocks = len(blocks)
+        data = self.dummy_cache[:2*num_blocks]
+        dummy_key_cpu = data[:num_blocks]
+        dummy_value_cpu = data[num_blocks:2*num_blocks]
+        with torch.cuda.stream(self.migration_stream):
+            for layer_idx in range(self.cache_engine.num_layers):
+                for idx, block_num in enumerate(blocks):
+                    dummy_key_cpu[idx][layer_idx].copy_(self.gpu_cache[layer_idx][0][block_num], non_blocking=True)
+                    dummy_value_cpu[idx][layer_idx].copy_(self.gpu_cache[layer_idx][1][block_num], non_blocking=True)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+        col.send(data, dst_handle, self.group_name)
+
+    def do_recv(self, src_handle, blocks: List[int]):
+        num_blocks = len(blocks)
+        data = self.dummy_cache[:2*num_blocks]
+        # note that col.recv use ray.collective inner stream, not migration_stream
+        col.recv(data, src_handle, self.group_name)
+
+        dummy_key_cpu = data[:num_blocks]
+        dummy_value_cpu = data[num_blocks:2*num_blocks]
+        with torch.cuda.stream(self.migration_stream):
+            for layer_idx in range(self.cache_engine.num_layers):
+                for idx, block_num in enumerate(blocks):
+                    self.gpu_cache[layer_idx][0][block_num].copy_(dummy_key_cpu[idx][layer_idx], non_blocking=True)
+                    self.gpu_cache[layer_idx][1][block_num].copy_(dummy_value_cpu[idx][layer_idx], non_blocking=True)
+        torch.cuda.Stream.synchronize(self.migration_stream)
+
+def get_migrate_backend(migrate_config: MigrationConfig, cache_engine: CacheEngine, worker_handle_list, scheduling_strategy, \
+                           dtype, is_driver_worker, gpu_cache, ray_world_size, ray_rank, worker_rank, local_rank) -> MigrateBackendBase:
+    if migrate_config.pp_or_tp_enabled and migrate_config.migration_backend == 'nccl':
+        logger.warning("NCCL backend is not supported for PP or TP enabled model, using gloo instead.")
+        migrate_config.migration_backend = 'gloo'
```

Comment:
May be we can directly detect it in the argument checker? We can add a argument checker in a seperate PR.

Reply:
yes